File: java/ofdpdfsigner_jni/src/main/java/com/fzz/seal/PDFStamp.java
package com.fzz.seal;

import com.fzz.seal.callback.ISignCallBack;
import com.fzz.seal.callback.IVerifyCallBack;
import com.fzz.seal.jni.IJNI;
import com.fzz.seal.jni.ReturnData;

import java.nio.charset.StandardCharsets;

public class PDFStamp {
    public static final int RIDING_LEFT = 0;
    public static final int RIDING_RIGHT = 1;
    public static final int RIDING_TOP = 2;
    public static final int RIDING_BOTTOM = 3;
    public static final int RIDING_CUTTING_RIGHT = 4;
    public static final int RIDING_CUTTING_TOP = 5;
    /*
     PDF 位置盖章
     imageData：印章数据
     pdfData：PDF数据
     pageIndex：页码 0开始
     x：开始位置
     y：开始位置
     stampwidth：印章宽
     stampheight: 印章高
     signCallBack：签名回调
     返回：电子印章数据
     */
//    public static byte[] signStamp_Pos(byte[] imageData,byte[] pdfData,double alpha, int pageIndex,double x,double y,
//                                       double stampwidth, double stampheight,
//                                       ISignCallBack signCallBack)
//    {
//        byte[] ret = null;
//        if ( imageData == null || imageData.length <= 0 || pdfData == null || pdfData.length <= 0 ||
//                stampwidth < 0 || stampheight < 0 || pageIndex < 0 ) {
//            return ret;
//        }
//        ReturnData returnData = new ReturnData();
//
//        IJNI.PDFSignStamp_Pos(imageData, pdfData, alpha, pageIndex, x, y,stampwidth,  stampheight,signCallBack,returnData);
//
//        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
//            byte[] temp = returnData.getResultData();
//            //System.out.println("zip len : "+temp.length);
//            ret = new byte[temp.length];
//            System.arraycopy(temp,0,ret,0,temp.length);
//        }
//        return ret;
//    }

    /*
     PDF 位置盖章
     imageData：印章数据
     pdfData：PDF数据
     pageIndex：页码 0开始
     x：开始位置
     y：开始位置
     stampwidth：印章宽
     stampheight: 印章高
     signCallBack：签名回调
     返回：电子印章数据
     */
    public static byte[] positionSign(long fileHandle,byte[] sealImageData,double sealAlpha, int pageNumber,double stampX,double stampY,
                                       double width, double height,boolean isLocked,
                                       ISignCallBack signCallBack)
    {
        byte[] ret = null;
        if ( sealImageData == null || sealImageData.length <= 0 || fileHandle == 0 ||
                width < 0 || height < 0 || pageNumber == 0  || pageNumber < -1 ) {
            return ret;
        }
        ReturnData returnData = new ReturnData();

        IJNI.PDFSignStamp_Pos_Handle(fileHandle,sealImageData, sealAlpha, pageNumber, stampX, stampY,width,  height,signCallBack,returnData);

        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
            byte[] temp = returnData.getResultData();
            //System.out.println("positionSign len : "+temp.length);
            ret = new byte[temp.length];
            System.arraycopy(temp,0,ret,0,temp.length);
        } else {
            //System.out.println("positionSign error:"+returnData.getErrorCode());
        }
        return ret;
    }
    /*
     PDF文件骑缝章签章
     imageData：印章数据
     pdfData：PDF数据
     RidingType：骑缝章类型
     stampwidth：印章宽
     stampheight: 印章高
     signCallBack：签名回调
     返回：电子印章数据
     */
//    public static byte[] signStamp_Riding(byte[] imageData,byte[] pdfData,double alpha, int RidingType,
//                                              double stampwidth, double stampheight,ISignCallBack signCallBack)
//    {
//        byte[] ret = null;
//        if ( imageData == null || imageData.length <= 0 || pdfData == null || pdfData.length <= 0 ||
//                stampwidth < 0 || stampheight < 0  ) {
//            return ret;
//        }
//        ReturnData returnData = new ReturnData();
//
//        IJNI.PDFSignStamp_Riding(imageData, pdfData, alpha,RidingType,stampwidth,  stampheight,signCallBack,returnData);
//
//        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
//            byte[] temp = returnData.getResultData();
//            //System.out.println("zip len : "+temp.length);
//            ret = new byte[temp.length];
//            System.arraycopy(temp,0,ret,0,temp.length);
//        }
//        return ret;
//    }

    /*
     PDF文件骑缝章签章
     imageData：印章数据
     pdfData：PDF数据
     RidingType：骑缝章类型
     stampwidth：印章宽
     stampheight: 印章高
     signCallBack：签名回调
     返回：电子印章数据
     */
    public static byte[] ridingSign(long fileHandle,byte[] sealImageData,double sealAlpha, String ridingType,
                                          double width, double height,boolean isLocked,ISignCallBack signCallBack)
    {
        byte[] ret = null;
        if ( sealImageData == null || sealImageData.length <= 0 || fileHandle == 0 ||
                width < 0 || height < 0  ) {
            return ret;
        }
        ReturnData returnData = new ReturnData();

        int iType = Integer.parseInt(ridingType);

        IJNI.PDFSignStamp_Riding_Handle(fileHandle,sealImageData, sealAlpha,iType,width,  height,signCallBack,returnData);

        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
            byte[] temp = returnData.getResultData();
            //System.out.println("zip len : "+temp.length);
            ret = new byte[temp.length];
            System.arraycopy(temp,0,ret,0,temp.length);
        }
        return ret;
    }
    /*
     PDF文件关键字签章
     imageData：印章数据
     pdfData：PDF数据
     keyword：关键字
     KeywordRule：关键字规则
     stampwidth：印章宽
     stampheight: 印章高
     signCallBack：签名回调
     返回：电子印章数据
     */
//    public static byte[] signStamp_KeyWord(byte[] imageData, byte[] pdfData, double alpha,String keyword,String KeywordRule,
//                                               double stampwidth, double stampheight,ISignCallBack signCallBack)
//    {
//        byte[] ret = null;
//        if ( imageData == null || imageData.length <= 0 || pdfData == null || pdfData.length <= 0 ||
//                stampwidth < 0 || stampheight < 0 || keyword == null || keyword.length() <= 0
//                || KeywordRule == null || KeywordRule.length() <= 0 ) {
//            return ret;
//        }
//        ReturnData returnData = new ReturnData();
//
//        IJNI.PDFSignStamp_KeyWord(imageData, pdfData, alpha, keyword.getBytes(StandardCharsets.UTF_8), KeywordRule.getBytes(StandardCharsets.UTF_8), stampwidth,  stampheight,signCallBack,returnData);
//
//        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
//            byte[] temp = returnData.getResultData();
//            //System.out.println("zip len : "+temp.length);
//            ret = new byte[temp.length];
//            System.arraycopy(temp,0,ret,0,temp.length);
//        }
//        return ret;
//    }

    /*
     PDF文件关键字签章
     imageData：印章数据
     pdfData：PDF数据
     keyword：关键字
     KeywordRule：关键字规则
     stampwidth：印章宽
     stampheight: 印章高
     signCallBack：签名回调
     返回：电子印章数据
     */
    public static byte[] keywordSign(long fileHandle,byte[] sealImageData, double sealAlpha,String keyWord,String keyWordRules,
                                           double width, double height,boolean isLocked,ISignCallBack signCallBack)
    {
        byte[] ret = null;
        if ( sealImageData == null || sealImageData.length <= 0 || fileHandle == 0 ||
                width < 0 || height < 0 || keyWord == null || keyWord.length() <= 0
                || keyWordRules == null || keyWordRules.length() <= 0 ) {
            return ret;
        }
        ReturnData returnData = new ReturnData();

        IJNI.PDFSignStamp_KeyWord_Handle(fileHandle,sealImageData, sealAlpha, keyWord.getBytes(StandardCharsets.UTF_8), keyWordRules.getBytes(StandardCharsets.UTF_8), width,  height,signCallBack,returnData);

        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
            byte[] temp = returnData.getResultData();
            //System.out.println("zip len : "+temp.length);
            ret = new byte[temp.length];
            System.arraycopy(temp,0,ret,0,temp.length);
        }
        return ret;
    }

    /*
     PDF 文档验证
     fileData：PDF数据
     passWord: 密码； "" 或 非""
     IVerifyCallBack：验证回调
     返回：验证结果，json数据
     */
    public static String verify(byte[] fileData, String passWord, IVerifyCallBack verifyCallBack)
    {
        String ret = "";
        if ( fileData == null || fileData.length <= 0 ) {
            return ret;
        }

        if ( passWord == null ) {
            passWord = "";
        }

        ReturnData returnData = new ReturnData();

        IJNI.PDFVerifyDocment(fileData,passWord.getBytes(StandardCharsets.UTF_8),verifyCallBack,returnData);

        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
            byte[] temp = returnData.getResultData();
            ret = new String(temp);
        }
        return ret;
    }

    /*
     PDF 删除文档中全部签名数据
     fileHandle：pdf 文档句柄
     返回：删除后文档数据
     */
    public static byte[] clearSigns(long fileHandle)
    {
        byte[] ret = null;
        if ( fileHandle == 0 ) {
            return ret;
        }
        ReturnData returnData = new ReturnData();

        IJNI.PDFClearStamps(fileHandle,returnData);

        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
            byte[] temp = returnData.getResultData();
            //System.out.println("zip len : "+temp.length);
            ret = new byte[temp.length];
            System.arraycopy(temp,0,ret,0,temp.length);
        }
        return ret;
    }

}

File: ios/ofdpdfsigner_ios/ofdpdfsignerlib/PDFStamp.swift
//
//  PDFStamp.swift
//  ofdcreatorlib
//
//  Created by zizheng fan on 2024/6/7.
//

import Foundation

//----------------------------------------------------------------------------------------
typealias Type_PDFSignStamp_Pos_Handle = @convention(c) (UnsafeMutableRawPointer,UnsafeRawPointer?, Int32,Double,Int32,Double,Double,Double,Double, Type_CallBack_Sign,Type_CallBack_getUserCert,UnsafeRawPointer?, UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?, UnsafeMutablePointer<size_t>?) -> Int32
private var Function_PDFSignStamp_Pos_Handle: Type_PDFSignStamp_Pos_Handle? = unsafeBitCast(dlsym(UnsafeMutableRawPointer(bitPattern: -2), "pdf_positionSign"), to: Type_PDFSignStamp_Pos_Handle.self)


public typealias Type_VerifyDocment = @convention(c) (UnsafeRawPointer?, Int32,Int32,UnsafePointer<CChar>?,Type_CallBack_Verify,UnsafeRawPointer?, UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?, UnsafeMutablePointer<size_t>?) -> Int32
public var Function_VerifyDocment: Type_VerifyDocment? = unsafeBitCast(dlsym(UnsafeMutableRawPointer(bitPattern: -2), "verifyDocment"), to: Type_VerifyDocment.self)

typealias Type_PDFSignStamp_Riding_Handle = @convention(c) (UnsafeMutableRawPointer,UnsafeRawPointer?, Int32,Double,Int32,Double,Double, Type_CallBack_Sign,Type_CallBack_getUserCert,UnsafeRawPointer?, UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?, UnsafeMutablePointer<size_t>?) -> Int32
private var Function_PDFSignStamp_Riding_Handle: Type_PDFSignStamp_Riding_Handle? = unsafeBitCast(dlsym(UnsafeMutableRawPointer(bitPattern: -2), "pdf_ridingSign"), to: Type_PDFSignStamp_Riding_Handle.self)

typealias Type_PDFSignStamp_KeyWord_Handle = @convention(c) (UnsafeMutableRawPointer,UnsafeRawPointer?, Int32,Double,UnsafePointer<CChar>?,UnsafePointer<CChar>?,Double,Double, Type_CallBack_Sign,Type_CallBack_getUserCert,UnsafeRawPointer?, UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?, UnsafeMutablePointer<size_t>?) -> Int32
private var Function_PDFSignStamp_KeyWord_Handle: Type_PDFSignStamp_KeyWord_Handle? = unsafeBitCast(dlsym(UnsafeMutableRawPointer(bitPattern: -2), "pdf_keywordSign"), to: Type_PDFSignStamp_KeyWord_Handle.self)


typealias Type_PDFSignStamp_ClearSigns_Handle = @convention(c) (UnsafeMutableRawPointer,UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?, UnsafeMutablePointer<size_t>?) -> Int32
private var Function_PDFSignStamp_ClearSigns_Handle: Type_PDFSignStamp_ClearSigns_Handle? = unsafeBitCast(dlsym(UnsafeMutableRawPointer(bitPattern: -2), "pdf_clearSigns"), to: Type_PDFSignStamp_ClearSigns_Handle.self)

//----------------------------------------------------------------------------------------
public class PDFStamp {
    
    public static let RIDING_LEFT : Int32 = 0;
    public static let RIDING_RIGHT : Int32 = 1;
    public static let RIDING_TOP : Int32 = 2;
    public static let RIDING_BOTTOM : Int32 = 3;
    public static let RIDING_CUTTING_RIGHT : Int32 = 4;
    public static let RIDING_CUTTING_TOP : Int32 = 5;
    /*
     PDF 位置盖章
     fileHandle:PDF文件句柄
     imageData：印章数据
     sealAlpha: 印章图片透明度
     pageIndex：页码 0开始
     x：开始位置
     y：开始位置
     stampwidth：印章宽
     stampheight: 印章高
     isLocked：是否继续盖章
     signCallback：签名回调
     getUserCertCallback:用户证书
     返回：电子印章数据
     */
    public static func positionSign(fileHandle : UInt,sealImageData : [UInt8]?,sealAlpha : Double, pageNumber : Int32,stampX : Double,stampY : Double,
                                    width : Double, height : Double,isLocked : Bool,
                                    withClosure signCallback: @escaping Type_CallBack_Sign,withClosure getUserCertCallback: @escaping Type_CallBack_getUserCert) -> [UInt8]?
    {
        var retBytes : [UInt8]? = nil;
        if fileHandle == 0 {
            return retBytes;
        }
        if sealImageData == nil || sealImageData!.count == 0 {
            return retBytes;
        }
        if let handle = UnsafeMutableRawPointer(bitPattern: fileHandle) {
            if let function = Function_PDFSignStamp_Pos_Handle {
                //print("createSeal")
                let inDataPointer = sealImageData!.withUnsafeBytes({ $0 }).baseAddress
                let inDatalen = Int32(sealImageData!.count)
                
                var dataPtr: UnsafeMutablePointer<UInt8>? = nil
                var datalen: size_t = 0
                
                let result = function(handle,inDataPointer,inDatalen,sealAlpha, pageNumber, stampX, stampY,width,height, signCallback,getUserCertCallback,nil,&dataPtr, &datalen)
                
                if result == 0 {
                    retBytes = Utils.convertUnsafePointerToArray(dataPtr: dataPtr, datalen: datalen)
                } else {
                    print("positionSign error:")
                    print(result)
                }
                
                if let pdataPtr = dataPtr {
                    Utils.freeHandle(handlePtr: pdataPtr)
                }
            }
        }

        return retBytes;
    }
    
    /*
     PDF 文档验证
     fileData：PDF数据
     verifyCallback：验证回调
     返回：验证结果，json数据
     */
    public static func verify(fileData : [UInt8]?,  password :String,withClosure verifyCallback: @escaping Type_CallBack_Verify) -> String?
    {
        var ret : String = ""
        if fileData == nil || fileData!.count == 0 {
            return ret;
        }
        
        if let function = Function_VerifyDocment {
            //print("verifySeal")
            let inDataPointer = fileData!.withUnsafeBytes({ $0 }).baseAddress
            let inDatalen = Int32(fileData!.count)
            
            var dataPtr: UnsafeMutablePointer<UInt8>? = nil
            var datalen: size_t = 0
            
            let result = function(inDataPointer,inDatalen,2,password,verifyCallback,nil,&dataPtr, &datalen)
            
            if result == 0 {
                //let retBytes = Utils.convertUnsafePointerToArray(dataPtr: dataPtr, datalen: datalen)
                ret = String(data: Utils.convertToData(bytes: dataPtr, lenght: Int32(datalen))!, encoding: .utf8)!
            } else {
                print("verify error:")
                print(result)
            }
            
            if let pdataPtr = dataPtr {
                Utils.freeHandle(handlePtr: pdataPtr)
            }
            
        }
        
        return ret;
    }
    
    /*
     PDF文件骑缝章签章
     fileHandle:PDF文件句柄
     imageData：印章数据
     sealAlpha: 印章图片透明度
     RidingType：骑缝章类型
     stampwidth：印章宽
     stampheight: 印章高
     isLocked：是否继续盖章
     signCallback：签名回调
     getUserCertCallback:用户证书
     返回：电子印章数据
     */
    public static func ridingSign(fileHandle : UInt,sealImageData : [UInt8]?,sealAlpha : Double, ridingType : String,
                                          width : Double, height : Double,isLocked : Bool,withClosure signCallback: @escaping Type_CallBack_Sign,withClosure getUserCertCallback: @escaping Type_CallBack_getUserCert)  -> [UInt8]?
    {
        var retBytes : [UInt8]? = nil;
        if fileHandle == 0 {
            return retBytes;
        }
        if sealImageData == nil || sealImageData!.count == 0 {
            return retBytes;
        }
        
        if let handle = UnsafeMutableRawPointer(bitPattern: fileHandle) {
            if let function = Function_PDFSignStamp_Riding_Handle {
                //print("createSeal")
                let inDataPointer = sealImageData!.withUnsafeBytes({ $0 }).baseAddress
                let inDatalen = Int32(sealImageData!.count)
                
                var dataPtr: UnsafeMutablePointer<UInt8>? = nil
                var datalen: size_t = 0
                
                let iType = Int32(ridingType)!
                
                let result = function(handle,inDataPointer,inDatalen,sealAlpha, iType, width,height, signCallback,getUserCertCallback,nil,&dataPtr, &datalen)
                
                if result == 0 {
                    retBytes = Utils.convertUnsafePointerToArray(dataPtr: dataPtr, datalen: datalen)
                } else {
                    print("positionSign error:")
                    print(result)
                }
                
                if let pdataPtr = dataPtr {
                    Utils.freeHandle(handlePtr: pdataPtr)
                }
            }
        }

        return retBytes;
    }
    
    
    /*
     PDF文件关键字签章
     fileHandle:PDF文件句柄
     imageData：印章数据
     sealAlpha: 印章图片透明度
     keyword：关键字
     KeywordRule：关键字规则
     stampwidth：印章宽
     stampheight: 印章高
     isLocked：是否继续盖章
     signCallback：签名回调
     getUserCertCallback:用户证书
     返回：电子印章数据
     */
    public static func keywordSign(fileHandle : UInt,sealImageData : [UInt8]?,sealAlpha : Double, keyWord : String,keyWordRules : String,
                                   width : Double, height : Double,isLocked : Bool,withClosure signCallback: @escaping Type_CallBack_Sign,withClosure getUserCertCallback: @escaping Type_CallBack_getUserCert) -> [UInt8]?
    {
        var retBytes : [UInt8]? = nil;
        if fileHandle == 0 {
            return retBytes;
        }
        if sealImageData == nil || sealImageData!.count == 0 {
            return retBytes;
        }
        
        if let handle = UnsafeMutableRawPointer(bitPattern: fileHandle) {
            if let function = Function_PDFSignStamp_KeyWord_Handle {
                //print("createSeal")
                let inDataPointer = sealImageData!.withUnsafeBytes({ $0 }).baseAddress
                let inDatalen = Int32(sealImageData!.count)
                
                var dataPtr: UnsafeMutablePointer<UInt8>? = nil
                var datalen: size_t = 0
                
              
                
                let result = function(handle,inDataPointer,inDatalen,sealAlpha, keyWord,keyWordRules, width,height, signCallback,getUserCertCallback,nil,&dataPtr, &datalen)
                
                if result == 0 {
                    retBytes = Utils.convertUnsafePointerToArray(dataPtr: dataPtr, datalen: datalen)
                } else {
                    print("positionSign error:")
                    print(result)
                }
                
                if let pdataPtr = dataPtr {
                    Utils.freeHandle(handlePtr: pdataPtr)
                }
            }
        }

        return retBytes;
    }
    
    /*
     删除PDF文档中全部印章和数字签名
     fileHandle:PDF文件句柄
     返回：删除PDF文档中全部印章和数字签名后的文档数据
     */
    public static func clearSigns(fileHandle : UInt) -> [UInt8]?
    {
        var retBytes : [UInt8]? = nil;
        if fileHandle == 0 {
            return retBytes;
        }
        if let handle = UnsafeMutableRawPointer(bitPattern: fileHandle) {
            if let function = Function_PDFSignStamp_ClearSigns_Handle {
              
                var dataPtr: UnsafeMutablePointer<UInt8>? = nil
                var datalen: size_t = 0
                
                let result = function(handle,&dataPtr, &datalen)
                
                if result == 0 {
                    retBytes = Utils.convertUnsafePointerToArray(dataPtr: dataPtr, datalen: datalen)
                } else {
                    print("clearSigns error:")
                    print(result)
                }
                
                if let pdataPtr = dataPtr {
                    Utils.freeHandle(handlePtr: pdataPtr)
                }
            }
        }

        return retBytes;
    }
}
//----------------------------------------------------------------------------------------
